peernotifier/peernotifier.go

```diff
@@ -0,0 +1,94 @@
+package peernotifier
+
+import (
+	"sync"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/subscribe"
+)
+
+// PeerNotifier is a subsystem which observes peer offline and online events.
+// It takes subscriptions for its events, and whenever it observes a new event
+// it notifies its subscribers over the proper channel.
+type PeerNotifier struct {
+	started sync.Once
+	stopped sync.Once
+
+	ntfnServer *subscribe.Server
+}
+
+// PeerOnlineEvent represents a new event where a peer comes online.
+type PeerOnlineEvent struct {
+	// PubKey is the ecdsa.PubKey of the peer that came online.
+	PubKey *btcec.PublicKey
+}
+
+// PeerOfflineEvent represents a new event where a peer goes offline.
+type PeerOfflineEvent struct {
+	// PubKey is the ecdsa.PubKey of the peer that went offline.
+	PubKey *btcec.PublicKey
+}
+
+// New creates a new peer notifier which notifies clients of peer online
+// and offline events.
+func New() *PeerNotifier {
+	return &PeerNotifier{
+		ntfnServer: subscribe.NewServer(),
+	}
+}
+
+// Start starts the PeerNotifier's subscription server.
+func (p *PeerNotifier) Start() error {
+	var err error
+
+	p.started.Do(func() {
+		log.Info("PeerNotifier starting")
+		err = p.ntfnServer.Start()
+	})
+
+	return err
+}
+
+// Stop signals the notifier for a graceful shutdown.
+func (p *PeerNotifier) Stop() {
+	p.stopped.Do(func() {
+		log.Info("Stopping PeerNotifier")
+		p.ntfnServer.Stop()
+	})
+}
+
+// SubscribePeerEvents returns a subscribe.Client that will receive updates
+// any time the Server is informed of a peer event.
+func (p *PeerNotifier) SubscribePeerEvents() (*subscribe.Client, error) {
+	return p.ntfnServer.Subscribe()
+}
+
+// NotifyPeerOnline sends a peer online event to all clients subscribed to the
+// peer notifier.
+func (p *PeerNotifier) NotifyPeerOnline(pubKey *btcec.PublicKey) {
+	event := PeerOnlineEvent{PubKey: pubKey}
+
+	log.Debugf(
```

Comment:
style nit: For logging and other string formatting calls we usually do
```go
log.Debugf("PeerNotifier notifying peer: %x online",
		string(pubKey.SerializeCompressed()[:]))
```

Reply:
cool, so generally for string formatting, keep the string inline and then newlines for all the vars like you do for functions? 
```
log.Debugf("Long sentence taking up space: %x and two values: %v", 
                   firstVar, 
                   secondVar,
)
```

Or just wrap at 80
```
log.Debugf("Long sentence taking up space: %x and two values: %v", 
                   firstVar, secondVar)
```

Reply:
@carlaKC typically we shoot for the latter wrt to logging statements